File: protocol-generator/typescript/src/peer.ts
import { readFileSync } from "node:fs";
import * as net from "node:net";
import { PeerIdentity } from "./identity/index.js";
import { ContentStore, EntityTree } from "./store/index.js";
import {
  CapabilityHandler,
  ConnectHandler,
  ConnectionState,
  type ExpressionEvaluator,
  type Handler,
  type HandlerBody,
  type HandlerHandle,
  HandlerRegistry,
  type HandlerSpec,
  HandlersHandler,
  type PeerServices,
  TreeHandler,
  ValidateEchoHandler,
  ValidateDispatchOutboundHandler,
} from "./handlers/index.js";
import { CapabilityToken, type GrantEntry, SeedPolicy, SeedPolicyError } from "./capability/index.js";
import { EmitBus } from "./emit/index.js";
import { Entity, Ecf, TypeNames, hashHex } from "./model/index.js";
import { Dispatcher } from "./dispatch/index.js";
import { seedCoreTypes } from "./types/index.js";
import { DEFAULT_MAX_FRAME_BYTES, PeerConnection, type PeerSession, initiate } from "./transport/index.js";

const HANDSHAKE_TIMEOUT_MS = 10_000;

/**
 * An Entity Core protocol peer (V7 Layers 0–4): identity, the codec, the entity
 * tree + content store, the bootstrap system handlers, the dispatch chain, and TCP
 * transport. Listens for inbound connections and dials outbound ones; both
 * directions complete the §4.1 handshake. No standard extensions are bundled —
 * community handlers register above the dispatcher boundary.
 */
export class Peer implements PeerServices {
  readonly #identity: PeerIdentity;
  readonly #emit: EmitBus;
  readonly #tree: EntityTree;
  readonly #store: ContentStore;
  readonly #registry: HandlerRegistry;
  readonly #dispatcher: Dispatcher;
  readonly #connections = new Set<PeerConnection>();
  readonly #seedPolicy: SeedPolicy;
  readonly #maxFrameBytes: number;
  #server: net.Server | null = null;
  #port = 0;

  /**
   * @param options.identity Peer identity; a fresh Ed25519 keypair is generated when omitted.
   * @param options.seedPolicy The §6.9a identity → capability seed policy materialized at
   * L0 and consulted at §4.6 authenticate (the `withSeedPolicy` builder affordance).
   * Defaults to the conformant {@link SeedPolicy.standard} (default → §4.4 discovery floor),
   * or — when `debugOpenGrants` is set — {@link SeedPolicy.debugOpen} (default → `*`).
   * @param options.debugOpenGrants Debug-only: select the degenerate `default → *` seed
   * policy (the retired `--debug-open-grants` behaviour, now routed through the real §6.9a
   * mechanism). Deprecated in v7.74, removed in v7.75. Ignored when `seedPolicy` is supplied.
   */
  constructor(
    options: {
      identity?: PeerIdentity;
      seedPolicy?: SeedPolicy;
      debugOpenGrants?: boolean;
      /**
       * Conformance-build opt-in (GUIDE-CONFORMANCE §7a): register the `system/validate/*`
       * test-handlers (`echo` + `dispatch-outbound`) so a black-box validator can drive the
       * §6.13(a)/(b) extensibility hooks. Conformance scaffolding, **not core protocol**, and
       * **off by default** — `dispatch-outbound` is an outbound originator that must never be
       * live in production. Surfaced as the host `--validate` switch.
       */
      conformanceHandlers?: boolean;
      /**
       * The frame budget (§1.6) applied to connections this peer opens or accepts, in
       * bytes. Defaults to {@link DEFAULT_MAX_FRAME_BYTES}. Surfaced to handler bodies
       * through `HandlerContext.frameBudget()` so a body sizes its response against the
       * limit its response is actually measured against.
       */
      maxFrameBytes?: number;
    } = {},
  ) {
    this.#identity = options.identity ?? PeerIdentity.generate();
    this.#maxFrameBytes = options.maxFrameBytes ?? DEFAULT_MAX_FRAME_BYTES;
    this.#seedPolicy = options.seedPolicy ?? (options.debugOpenGrants ? SeedPolicy.debugOpen() : SeedPolicy.standard());
    this.#emit = new EmitBus();
    this.#store = new ContentStore(this.#emit);
    this.#tree = new EntityTree(this.#store, this.#emit);
    this.#registry = new HandlerRegistry(this);
    this.#dispatcher = new Dispatcher(this, this.#registry);
    this.#bootstrap();
    if (options.conformanceHandlers) {
      // §7a.2: registered only on the conformance opt-in; never in a default/production peer.
      this.registerHandler(new ValidateEchoHandler());
      this.registerHandler(new ValidateDispatchOutboundHandler());
    }
  }

  // ----- PeerServices ---------------------------------------------------

  get localPeerId(): string {
    return this.#identity.peerId;
  }

  get localIdentity(): PeerIdentity {
    return this.#identity;
  }

  get tree(): EntityTree {
    return this.#tree;
  }

  get contentStore(): ContentStore {
    return this.#store;
  }

  get emit(): EmitBus {
    return this.#emit;
  }

  get nowMs(): bigint {
    return BigInt(Date.now());
  }

  /** The peer's configured default frame budget (§1.6), in bytes. */
  get maxFrameBytes(): number {
    return this.#maxFrameBytes;
  }

  /** The port the listener is bound to (valid after {@link listen}). */
  get port(): number {
    return this.#port;
  }

  // ----- lifecycle ------------------------------------------------------

  #bootstrap(): void {
    // Bootstrap handlers (§6.9): tree + connect + handlers + capability are MUST. The
    // handlers handler (§6.2 / §6.13(a)) executes register/unregister behaviorally — a 501
    // stub is non-conformant. (Types handler SHOULD remains A-007.)
    this.#registry.register(new ConnectHandler());
    this.#registry.register(new TreeHandler());
    this.#registry.register(new HandlersHandler());
    this.#registry.register(new CapabilityHandler());

    // Local peer entity at system/peer/self (§3.13), tree-walkable.
    this.#tree.put("/" + this.localPeerId + "/system/peer/self", this.#identity.peerEntity);

    // Core type registry → system/type/* (TYPE-SYSTEM §8–§10). Core + operational +
    // type-system bootstrap only (53 types; refined G4 / F17).
    seedCoreTypes(this.#tree, this.localPeerId);

    // §6.9a Peer Authority Bootstrap: materialize the seed capability entities into the
    // tree at L0 — the self-owner cap plus the seed-policy entries authenticate reads back.
    this.#seedAuthorityBootstrap();
  }

  /**
   * §6.9a Bootstrap L0 write-set (item 4): materialize the seed capability entities — the
   * `self`-owner capability (a root cap, full scope over `/{peer_id}/*`, grantee = the
   * peer's own identity, in the §6.9a.0 detached-signature shape: the cap token at the hex
   * policy path, its self-signature at the §3.5 invariant pointer), the `default` scope
   * template at the sentinel path, and any explicitly-named entries (§6.9a.1). Read back by
   * {@link ConnectHandler} at §4.6 authenticate via the v7.64 dual-form lookup.
   */
  #seedAuthorityBootstrap(): void {
    const policyBase = "/" + this.localPeerId + "/system/capability/policy/";

    // (1) self-owner capability (§6.9a.0 shape 1 — detached-signature).
    const { token: ownerCap, signature: ownerSig } = CapabilityToken.createRoot(
      this.#identity,
      this.#identity.identityHash,
      SeedPolicy.ownerGrants(this.localPeerId),
      this.nowMs,
    );
    this.#tree.put(policyBase + hashHex(this.#identity.identityHash), ownerCap.entity);
    this.#tree.put("/" + this.localPeerId + "/system/signature/" + ownerCap.contentHashHex, ownerSig);

    // (2) default seed entry — the fallback scope for any other authenticated identity.
    this.#tree.put(policyBase + "default", policyEntryEntity("default", this.#seedPolicy.defaultGrants));

    // (3) explicitly-named operator/admin/reader entries.
    for (const entry of this.#seedPolicy.namedEntries) {
      this.#tree.put(policyBase + entry.key, policyEntryEntity(entry.key, entry.grants));
    }
  }

  /**
   * `SDK-OPERATIONS` §11.6 — install a language-native body behind a {@link HandlerSpec}
   * and return its {@link HandlerHandle} (keystone peer contract `install.handler`,
   * `install.remove`, `install.grant`, `install.types`). This is the certified
   * registration surface.
   *
   * Performs the core §6.13(a) writes (types, `system/handler` entity, the handler's
   * grant minted with `spec.internalScope` — a grant covering nothing when null — its
   * signature, the interface) and binds the body in the dispatch index. Throws
   * {@link RegisterError} before writing anything: `409 pattern_collision` for a bound
   * pattern (including a bootstrap one), `400 invalid_handler_spec` for an invalid spec.
   * Does not refuse `system/*`. The body receives a {@link DispatchContext}, which only
   * the dispatcher constructs.
   */
  registerHandler(spec: HandlerSpec, body: HandlerBody): HandlerHandle;
  /**
   * Install a native (in-process) handler post-bootstrap — the seam an SDK / native
   * extension uses to add a handler with a compiled body (complementing the wire
   * `register` of §6.13(a), which installs entity-native bodies).
   *
   * The pre-contract surface, kept as it was: it replaces whatever is installed at the
   * pattern, refuses nothing, writes an empty-scope grant, and hands the body a
   * {@link HandlerContext} anyone can construct. **Not a keystone peer contract binding,
   * and not certified** — build on `registerHandler(spec, body)`.
   */
  registerHandler(handler: Handler): void;
  registerHandler(handlerOrSpec: Handler | HandlerSpec, body?: HandlerBody): HandlerHandle | void {
    if (typeof body === "function") {
      return this.#registry.install(handlerOrSpec as HandlerSpec, body);
    }
    this.#registry.register(handlerOrSpec as Handler);
  }

  /**
   * Install (or clear, with `null`) the evaluator for §6.13(a) entity-native handler
   * bodies — the seam a compute extension occupies.
   *
   * The built-in `compute/literal` fast path is unaffected and still answers first; an
   * installed evaluator receives only the bodies the core peer would otherwise refuse
   * with `501 unsupported_expression`, and may itself decline (`null`) to leave that
   * `501` in place. A peer with no evaluator installed behaves exactly as before.
   */
  setExpressionEvaluator(evaluator: ExpressionEvaluator | null): void {
    this.#dispatcher.setExpressionEvaluator(evaluator);
  }

  /** The installed §6.13(a) body evaluator, or `null` when none is installed. */
  get expressionEvaluator(): ExpressionEvaluator | null {
    return this.#dispatcher.expressionEvaluator;
  }

  /**
   * Begin listening at `host:port` (0 = auto-assign; `host` defaults to loopback).
   * Resolves with the bound port.
   */
  listen(port = 0, host = "127.0.0.1"): Promise<number> {
    return new Promise<number>((resolve, reject) => {
      const server = net.createServer((socket) => this.#onInbound(socket));
      server.once("error", reject);
      server.listen(port, host, () => {
        const address = server.address();
        this.#port = typeof address === "object" && address !== null ? address.port : port;
        this.#server = server;
        resolve(this.#port);
      });
    });
  }

  #onInbound(socket: net.Socket): void {
    socket.setNoDelay(true);
    const state = new ConnectionState();
    const conn = new PeerConnection(socket, this.#dispatcher, state, this.#maxFrameBytes);
    this.#connections.add(conn);
    conn.start();

    // No reverse-direction handshake here (§4.1 leg 3). The spec pins leg 3 as
    // OPTIONAL and reachability-gated, deferred until a signaling mechanism (a
    // `hello` "accepts-inbound" capability or similar) is designed: "A responder
    // MUST NOT proactively send a leg-3 authenticate to an initiator that has
    // not indicated it accepts inbound dispatch... An unsolicited inbound
    // authenticate corrupts a client-style initiator's next read." (§4.1; "no
    // reference impl currently sends leg 3"). This peer previously fired it
    // unconditionally on every inbound connection — a real conformance bug, not
    // a cosmetic one: it is the confirmed root cause of the
    // connectivity/handshake_nonce_single_use (RT-6) failure, where this eager
    // leg-3 EXECUTE reliably outraced a same-connection follow-up request (the
    // RT-6 replay probe) and got read by the peer on the other end as if it
    // were that request's response, desyncing the exchange and causing it to
    // disconnect before the real response was written.
  }

  /** Dial a peer at `host:port` and complete the handshake, returning the session. */
  connect(host: string, port: number, timeoutMs = HANDSHAKE_TIMEOUT_MS): Promise<PeerSession> {
    return new Promise<PeerSession>((resolve, reject) => {
      const socket = net.connect({ host, port }, () => {
        socket.setNoDelay(true);
        const state = new ConnectionState();
        const conn = new PeerConnection(socket, this.#dispatcher, state, this.#maxFrameBytes);
        this.#connections.add(conn);
        conn.start();
        initiate(conn, this.#identity, state, timeoutMs).then(resolve, reject);
      });
      socket.once("error", reject);
    });
  }

  /**
   * Stop listening and close every connection. The listening socket is released FIRST
   * and synchronously: `server.close` stops accepting at once, but its callback waits for
   * open connections to end, so awaiting it before destroying them could hold the process
   * (and never release the peer) for as long as a client keeps a connection open.
   */
  async dispose(): Promise<void> {
    let closed: Promise<void> = Promise.resolve();
    if (this.#server !== null) {
      const server = this.#server;
      this.#server = null;
      closed = new Promise<void>((resolve) => server.close(() => resolve()));
    }
    for (const conn of this.#connections) {
      await conn.dispose();
    }
    this.#connections.clear();
    await closed;
  }
}

/**
 * Load a keystone seed-policy file (`--seed-policy <path>`; convention README §5
 * `with_seed_policy_from_file`) and return it as a {@link Peer} options fragment:
 *
 *     new Peer({ identity, ...withSeedPolicyFromFile("policy.json") })
 *
 * Parsing and every refusal are {@link SeedPolicy.fromJson}'s; this adds only the read,
 * and prefixes any error with the path. Lives in the Node layer (beside `node:net`) so
 * `capability/` stays free of `node:*` imports for the browser bundle.
 */
export function withSeedPolicyFromFile(path: string): { seedPolicy: SeedPolicy } {
  let text: string;
  try {
    // Fatal decode: invalid UTF-8 is refused, not replaced with U+FFFD.
    text = new TextDecoder("utf-8", { fatal: true }).decode(readFileSync(path));
  } catch (err) {
    throw new SeedPolicyError(`${path}: ${err instanceof Error ? err.message : String(err)}`);
  }
  try {
    return { seedPolicy: SeedPolicy.fromJson(text) };
  } catch (err) {
    throw new SeedPolicyError(`${path}: ${err instanceof Error ? err.message : String(err)}`);
  }
}

/** Build a `system/capability/policy-entry` carrying a scope template (§6.9a.0 shape 2 / v7.62 §4). */
function policyEntryEntity(peerPattern: string, grants: readonly GrantEntry[]): Entity {
  return Entity.create(
    TypeNames.CapabilityPolicyEntry,
    Ecf.map(["peer_pattern", Ecf.text(peerPattern)], ["grants", Ecf.array(grants.map((g) => g.toEcf()))]),
  );
}

File: protocol-generator/csharp/src/EntityCore.Protocol/Peer.cs
using System.Collections.Concurrent;
using System.Net;
using System.Net.Sockets;
using EntityCore.Protocol.Capability;
using EntityCore.Protocol.Codec;
using EntityCore.Protocol.Dispatch;
using EntityCore.Protocol.Handlers;
using EntityCore.Protocol.Identity;
using EntityCore.Protocol.Model;
using EntityCore.Protocol.Store;
using EntityCore.Protocol.Transport;
using EntityCore.Protocol.Types;

namespace EntityCore.Protocol;

/// <summary>
/// An Entity Core protocol peer (V7 Layers 0–4): identity, the codec, the entity
/// tree + content store, the bootstrap system handlers, the dispatch chain, and TCP
/// transport. Listens for inbound connections and dials outbound ones; both
/// directions complete the §4.1 handshake. No standard extensions are bundled —
/// community handlers register above the dispatcher boundary.
/// </summary>
internal sealed class Peer : IPeerServices, IAsyncDisposable
{
    private readonly PeerIdentity _identity;
    private readonly Emit.EmitBus _emit;
    private readonly EntityTree _tree;
    private readonly ContentStore _store;
    private readonly HandlerRegistry _registry;
    private readonly Dispatcher _dispatcher;
    private readonly CancellationTokenSource _shutdown = new();
    private readonly ConcurrentBag<PeerConnection> _connections = new();
    private TcpListener? _listener;
    private Task _acceptLoop = Task.CompletedTask;
    private readonly SeedPolicy _seedPolicy;

    /// <param name="identity">Peer identity; a fresh keypair is generated when null.</param>
    /// <param name="seedPolicy">
    /// The §6.9a identity → capability seed policy materialized at L0 and consulted at
    /// §4.6 authenticate (the <c>with_seed_policy</c> builder affordance). Defaults to
    /// the conformant <see cref="SeedPolicy.Standard"/> (default → §4.4 discovery floor),
    /// or — when <paramref name="debugOpenGrants"/> is set — the degenerate
    /// <see cref="SeedPolicy.DebugOpen"/> (default → <c>*</c>).
    /// </param>
    /// <param name="debugOpenGrants">
    /// Debug-only: select the degenerate <c>default → *</c> seed policy (the retired
    /// <c>--debug-open-grants</c> behaviour, now routed through the real §6.9a mechanism
    /// rather than a hardcoded fork). Deprecated in v7.74, removed in v7.75. Ignored when
    /// an explicit <paramref name="seedPolicy"/> is supplied. The conformant default is false.
    /// </param>
    /// <param name="conformanceHandlers">
    /// Conformance-build opt-in (GUIDE-CONFORMANCE §7a): register the <c>system/validate/*</c>
    /// test-handlers (<c>echo</c> + <c>dispatch-outbound</c>) so a black-box validator can
    /// drive the §6.13(a)/(b) extensibility hooks. These are conformance scaffolding, <b>not
    /// core protocol</b>, and <b>off by default</b> — <c>dispatch-outbound</c> is an outbound
    /// originator that must never be live in a production install. Surfaced as the host
    /// <c>--validate</c> switch. The conformant production default is false.
    /// </param>
    public Peer(PeerIdentity? identity = null, SeedPolicy? seedPolicy = null, bool debugOpenGrants = false,
        bool conformanceHandlers = false)
    {
        _identity = identity ?? PeerIdentity.Generate();
        _seedPolicy = seedPolicy ?? (debugOpenGrants ? SeedPolicy.DebugOpen() : SeedPolicy.Standard());
        _emit = new Emit.EmitBus();
        _store = new ContentStore(_emit);
        _tree = new EntityTree(_store, _emit);
        _registry = new HandlerRegistry(this);
        _dispatcher = new Dispatcher(this, _registry);
        Bootstrap();
        if (conformanceHandlers)
        {
            RegisterConformanceHandlers();
        }
    }

    /// <summary>
    /// GUIDE-CONFORMANCE §7a: register the two <c>system/validate/*</c> conformance
    /// test-handlers as native handlers (same install path as any native handler — the
    /// five normative writes). Called only on the conformance opt-in; never in a default /
    /// production peer (the §7a.2 "not in base install" requirement).
    /// </summary>
    private void RegisterConformanceHandlers()
    {
        _registry.Register(new EchoHandler());
        _registry.Register(new DispatchOutboundHandler());
    }

    // ----- IPeerServices --------------------------------------------------

    public string LocalPeerId => _identity.PeerId;

    public PeerIdentity LocalIdentity => _identity;

    public EntityTree Tree => _tree;

    public ContentStore ContentStore => _store;

    public Emit.EmitBus Emit => _emit;

    public ulong NowMs => (ulong)DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    /// <summary>The port the listener is bound to (valid after <see cref="ListenAsync"/>).</summary>
    public int Port { get; private set; }

    // ----- lifecycle ------------------------------------------------------

    private void Bootstrap()
    {
        // Bootstrap handlers (§6.9): tree + connect + handlers + capability are MUST.
        // The handlers handler (§6.2 / §6.13(a)) executes register/unregister behaviorally
        // — a 501 stub is non-conformant. (Types handler SHOULD remains A-007.)
        _registry.Register(new ConnectHandler());
        _registry.Register(new TreeHandler());
        _registry.Register(new HandlersHandler());
        _registry.Register(new CapabilityHandler());

        // Local peer entity at system/peer/self (§3.13), tree-walkable.
        _tree.Put("/" + LocalPeerId + "/system/peer/self", _identity.PeerEntity);

        // Core type registry → system/type/* (TYPE-SYSTEM §8–§10). Core + operational
        // + type-system bootstrap only (53 types; refined G4 / F17). Rendered natively
        // and proven byte-identical to the oracle's registry (CoreTypeRegistryTests).
        CoreTypeRegistry.Seed(_tree, LocalPeerId);

        // §6.9a Peer Authority Bootstrap: materialize the seed capability entities into
        // the tree at L0 (pre-capability) — the self-owner cap plus the seed-policy
        // entries the §4.6 authenticate lookup reads back.
        SeedAuthorityBootstrap();
    }

    /// <summary>
    /// §6.9a Bootstrap L0 write-set (item 4): materialize the seed capability entities.
    /// <list type="number">
    /// <item>The <c>self</c>-owner capability — a root cap, full scope over the local
    ///   namespace <c>/{peer_id}/*</c>, grantee = the peer's own identity, in the
    ///   §6.9a.0 detached-signature shape (keystone S8-uniform): the cap token at the
    ///   hex policy path, its self-signature at the §3.5 invariant pointer.</item>
    /// <item>The <c>default</c> seed entry — the fallback scope template for any other
    ///   authenticated identity, stored as a <c>policy-entry</c> at the sentinel path.</item>
    /// <item>Any explicitly-named operator / admin / reader entries (§6.9a.1).</item>
    /// </list>
    /// Read back by <see cref="Handlers.ConnectHandler"/> at §4.6 authenticate via the
    /// v7.64 dual-form lookup (hex → Base58 → <c>default</c>).
    /// </summary>
    private void SeedAuthorityBootstrap()
    {
        string policyBase = "/" + LocalPeerId + "/system/capability/policy/";

        // (1) self-owner capability (§6.9a.0 shape 1 — detached-signature).
        (CapabilityToken ownerCap, Entity ownerSig) = CapabilityToken.CreateRoot(
            _identity, _identity.IdentityHash, SeedPolicy.OwnerGrants(LocalPeerId), NowMs);
        _tree.Put(policyBase + Hashes.Hex(_identity.IdentityHash), ownerCap.Entity);
        _tree.Put("/" + LocalPeerId + "/system/signature/" + ownerCap.ContentHashHex, ownerSig);

        // (2) default seed entry — the fallback scope for any other authenticated identity.
        _tree.Put(policyBase + "default", PolicyEntryEntity("default", _seedPolicy.DefaultGrants));

        // (3) explicitly-named operator/admin/reader entries.
        foreach (SeedPolicyEntry entry in _seedPolicy.NamedEntries)
        {
            _tree.Put(policyBase + entry.Key, PolicyEntryEntity(entry.Key, entry.Grants));
        }
    }

    /// <summary>Build a <c>system/capability/policy-entry</c> carrying a scope template (§6.9a.0 shape 2 / v7.62 §4).</summary>
    private static Entity PolicyEntryEntity(string peerPattern, IReadOnlyList<GrantEntry> grants) =>
        Entity.Create(TypeNames.CapabilityPolicyEntry, Ecf.Map(
            ("peer_pattern", Ecf.Text(peerPattern)),
            ("grants", Ecf.Array(grants.Select(g => g.ToEcf())))));

    /// <summary>
    /// Install a native (in-process) handler post-bootstrap — the seam an SDK / native
    /// extension uses to add a handler with a compiled body (complementing the wire
    /// <c>register</c> of §6.13(a), which installs entity-native bodies). Writes the same
    /// manifest / interface / grant / grant-signature entities as a bootstrap handler.
    /// </summary>
    public void RegisterHandler(IHandler handler) => _registry.Register(handler);

    /// <summary>Begin listening on loopback at <paramref name="port"/> (0 = auto-assign).</summary>
    public void ListenAsync(int port)
    {
        _listener = new TcpListener(IPAddress.Loopback, port);
        _listener.Start();
        Port = ((IPEndPoint)_listener.LocalEndpoint).Port;
        _acceptLoop = Task.Run(() => AcceptLoopAsync(_shutdown.Token));
    }

    private async Task AcceptLoopAsync(CancellationToken ct)
    {
        try
        {
            while (!ct.IsCancellationRequested)
            {
                TcpClient client = await _listener!.AcceptTcpClientAsync(ct).ConfigureAwait(false);
                client.NoDelay = true;
                var state = new ConnectionState();
                var conn = new PeerConnection(client.GetStream(), _dispatcher, state);
                _connections.Add(conn);
                conn.Start();

                // Reverse-direction handshake (§4.1 E3): the responder sends its own
                // authenticate once it has the initiator's hello. Fire-and-forget —
                // it completes the mutual handshake; its session is not needed here.
                _ = RespondInBackgroundAsync(conn, state, ct);
            }
        }
        catch (Exception) when (ct.IsCancellationRequested)
        {
            // Clean shutdown.
        }
        catch (ObjectDisposedException)
        {
            // Listener stopped.
        }
    }

    private async Task RespondInBackgroundAsync(PeerConnection conn, ConnectionState state, CancellationToken ct)
    {
        try
        {
            await Handshake.RespondAsync(conn, _identity, state, TimeSpan.FromSeconds(10), ct).ConfigureAwait(false);
        }
        catch (Exception)
        {
            // The reverse handshake is best-effort; failures don't affect the
            // initiator's already-established session.
        }
    }

    /// <summary>Dial a peer at <paramref name="host"/>:<paramref name="port"/> and complete the handshake.</summary>
    public async Task<PeerSession> ConnectAsync(string host, int port, TimeSpan timeout, CancellationToken ct = default)
    {
        var client = new TcpClient { NoDelay = true };
        await client.ConnectAsync(host, port, ct).ConfigureAwait(false);
        var state = new ConnectionState();
        var conn = new PeerConnection(client.GetStream(), _dispatcher, state);
        _connections.Add(conn);
        conn.Start();
        return await Handshake.InitiateAsync(conn, _identity, state, timeout, ct).ConfigureAwait(false);
    }

    public async ValueTask DisposeAsync()
    {
        await _shutdown.CancelAsync().ConfigureAwait(false);
        _listener?.Stop();
        try
        {
            await _acceptLoop.ConfigureAwait(false);
        }
        catch
        {
            // Accept-loop teardown errors are expected during shutdown.
        }
        foreach (PeerConnection conn in _connections)
        {
            await conn.DisposeAsync().ConfigureAwait(false);
        }
        _shutdown.Dispose();
    }
}
